Cgen accesses `len` field of optional field inside generated `indent_X`

### Describe the bug

<title>

### Reproduction Steps

```v
import x.json2

pub fn maybe_map[T, X](a []T, f fn (T) !X) ![]X {
	mut r := []X{cap: a.len}
	for v in a {
		r << f(v)!
	}
	return r
}

pub enum ApplicationCommandOptionType {
	sub_command       = 1
	sub_command_group
	string
	integer
	boolean
	user
	channel
	role
	mentionable
	number
	attachment
}

pub struct ApplicationCommandOption {
pub:
	typ ApplicationCommandOptionType @[required]
	name string @[required]
	description string @[required]
	options ?[]ApplicationCommandOption
}

pub fn ApplicationCommandOption.parse(j json2.Any) !ApplicationCommandOption {
	match j {
		map[string]json2.Any {
			return ApplicationCommandOption{
				typ: unsafe { ApplicationCommandOptionType(j['type']!.int()) }
				name: j['name']! as string
				description: j['description']! as string
				options: if a := j['options'] {
					maybe_map(a as []json2.Any, fn (k json2.Any) !ApplicationCommandOption {
						return ApplicationCommandOption.parse(k)!
					})!
				} else {
					none
				}
			}
		}
		else {
			return error('expected application command option to be object, got ${j.type_name()}')
		}
	}
}

fn main() {
	assert ApplicationCommandOption.parse({
		'type': json2.Any(3)
		'name': 'foo'
		'description': 'This is my first command.'
	}) or {
		assert false, 'Should not return error: ${err}'
		return
	} == ApplicationCommandOption{typ: .string, name: 'foo', description: 'This is my first command.'}
}
```

### Expected Behavior

Successful compile, and execution

### Current Behavior

```v
==================
C:/Users/mclr/AppData/Local/Temp/v_0/test.5793022678155136607.tmp.c:523: warning: WINVER redefined
C:/Users/mclr/AppData/Local/Temp/v_0/test.5793022678155136607.tmp.c:3073: error: field not found: len
...
==================
(Use `v -cg` to print the entire error message)

builder error:
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose
```

### Possible Solution

_No response_

### Additional Information/Context

i'm making `discord.v` models coverage with tests

### V version

V 0.4.3 c0d7c58

### Environment details (OS name and version, etc.)

```
V 0.4.3 c0d7c58
V full version: V 0.4.3 57a33c3.c0d7c58
OS: windows, Microsoft Windows 11 Pro v22000 64-bit
Processor: 12 cpus, 64bit, little endian, 

getwd: D:\Games\Proekti\V\interactions
vexe: D:\Games\Proekti\V\v\v.exe
vexe mtime: 2024-01-02 14:05:19

vroot: OK, value: D:\Games\Proekti\V\v
VMODULES: OK, value: C:\Users\mclr\.vmodules
VTMP: OK, value: C:\Users\mclr\AppData\Local\Temp\v_0

Git version: git version 2.37.0.windows.1
Git vroot status: weekly.2023.52-50-gc0d7c58b
.git/config present: true

CC version:
thirdparty/tcc status: thirdparty-windows-amd64 a39eb79b
```
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/auto_eq_methods.v
+++ b/vlib/v/gen/c/auto_eq_methods.v
@@ -342,6 +342,7 @@ fn (mut g Gen) gen_array_equality_fn(left_type ast.Type) string {
 
 	if left_type.has_flag(.option) {
 		fn_builder.writeln('\tif (a.state != b.state) return false;')
+		fn_builder.writeln('\tif (a.state == 2 && a.state == b.state) return true;')
 	}
 
 	fn_builder.writeln('\tif (${left_len} != ${right_len}) {')

--- a/vlib/v/gen/c/auto_str_methods.v
+++ b/vlib/v/gen/c/auto_str_methods.v
@@ -1022,7 +1022,12 @@ fn (mut g Gen) gen_str_for_struct(info ast.Struct, lang ast.Language, styp strin
 		// handle circular ref type of struct to the struct itself
 		if styp == field_styp && !allow_circular {
 			if is_field_array {
-				fn_body.write_string('it.${field_name}.len > 0 ? ${funcprefix}_SLIT("[<circular>]") : ${funcprefix}_SLIT("[]")')
+				if is_opt_field {
+					arr_styp := g.base_type(field.typ)
+					fn_body.write_string('it.${field_name}.state != 2 && (*(${arr_styp}*)it.${field_name}.data).len > 0 ? ${funcprefix}_SLIT("[<circular>]") : ${funcprefix}_SLIT("[]")')
+				} else {
+					fn_body.write_string('it.${field_name}.len > 0 ? ${funcprefix}_SLIT("[<circular>]") : ${funcprefix}_SLIT("[]")')
+				}
 			} else {
 				fn_body.write_string('${funcprefix}_SLIT("<circular>")')
 			}

--- a/vlib/v/tests/option_arr_auto_str_test.v
+++ b/vlib/v/tests/option_arr_auto_str_test.v
@@ -57,7 +57,10 @@ fn test_main() {
 		'type':        json2.Any(3)
 		'name':        'foo'
 		'description': 'This is my first command.'
-	})! == ApplicationCommandOption{
+	}) or {
+		assert false, 'Should not return error: ${err}'
+		return
+	} == ApplicationCommandOption{
 		typ: .string
 		name: 'foo'
 		description: 'This is my first command.'

--- a/vlib/v/tests/option_arr_auto_str_test.v
+++ b/vlib/v/tests/option_arr_auto_str_test.v
@@ -1,0 +1,65 @@
+import x.json2
+
+pub fn maybe_map[T, X](a []T, f fn (T) !X) ![]X {
+	mut r := []X{cap: a.len}
+	for v in a {
+		r << f(v)!
+	}
+	return r
+}
+
+pub enum ApplicationCommandOptionType {
+	sub_command       = 1
+	sub_command_group
+	string
+	integer
+	boolean
+	user
+	channel
+	role
+	mentionable
+	number
+	attachment
+}
+
+pub struct ApplicationCommandOption {
+pub:
+	typ         ApplicationCommandOptionType @[required]
+	name        string                       @[required]
+	description string                       @[required]
+	options     ?[]ApplicationCommandOption
+}
+
+pub fn ApplicationCommandOption.parse(j json2.Any) !ApplicationCommandOption {
+	match j {
+		map[string]json2.Any {
+			return ApplicationCommandOption{
+				typ: unsafe { ApplicationCommandOptionType(j['type']!.int()) }
+				name: j['name']! as string
+				description: j['description']! as string
+				options: if a := j['options'] {
+					maybe_map(a as []json2.Any, fn (k json2.Any) !ApplicationCommandOption {
+						return ApplicationCommandOption.parse(k)!
+					})!
+				} else {
+					none
+				}
+			}
+		}
+		else {
+			return error('expected application command option to be object, got ${j.type_name()}')
+		}
+	}
+}
+
+fn test_main() {
+	assert ApplicationCommandOption.parse({
+		'type':        json2.Any(3)
+		'name':        'foo'
+		'description': 'This is my first command.'
+	})! == ApplicationCommandOption{
+		typ: .string
+		name: 'foo'
+		description: 'This is my first command.'
+	}
+}

